File: Alignment/MillePedeAlignmentAlgorithm/scripts/mps_setup.pl
#!/usr/bin/env perl
#     R. Mankel, DESY Hamburg     08-Oct-2007
#     A. Parenti, DESY Hamburg    16-Apr-2008
#     $Revision: 1.26 $
#     $Date: 2011/11/22 14:21:01 $
#
#  Setup local mps database
#  
#
#  Usage:
#
#  mps_setup.pl [options] batchScript cfgTemplate infiList nJobs class[:classMerge] jobname [mergeScript [pool:]mssDir]
#
# class can be - any of the normal LSF queues (8nm,1nh,8nh,1nd,2nd,1nw,2nw)
#              - special CAF queues (cmscaf1nh, cmscaf1nd, cmscaf1nw)
#              - special CAF queues, for pede job (cmscafspec1nh, cmscafspec1nd, cmscafspec1nw). E.g. cmscafspec1nh corresponds to "-q cmscaf1nh -R cmscafspec"
# If class contains a ':', it will be split:
#              - the part before the ':' defines the class for Mille jobs,
#              - the part behind the class for the Pede job.
#              (eg cmscaf:cmscafspec)
# If mssDir contains a ':', it will be split:
#              - the part before ':' defines the pool,
#              - the part behind it the directory.
#              (eg cmscafuser:/castor/cern.ch/cms/store/...)
# Known options:
#  -m          Setup pede merging job.
#  -a          Append jobs to existing list.
#  -M pedeMem  The memory (MB) to be allocated for pede (min: 1024 MB).
#              If not given, it is evinced from the pede executable name.
#              Finally, it is set 2560 MB if neither of the two are available.
#  -N name     The name to be assigned to the jobs. Whitespaces and colons are not allowed.
#  -w          Assign statistical weight.

BEGIN {
use File::Basename;
unshift(@INC, dirname($0)."/mpslib");
}
use Mpslib;

$batchScript = "undefined";
$cfgTemplate = "undefined";
$infiList = "undefined";
$nJobs = 0;
$class = "S";
$addFiles = "";
$driver = "";
$mergeScript = "";
$mssDirPool = "";
$mssDir = "";
$append = 0;

my $confname = "";
my $weight = 1.0;
my $useweight = 0;

# parse the arguments
while (@ARGV) {
  my $arg = shift(ARGV);
  if ($arg =~ /\A-/) {  # check for option
    if ($arg =~ /-N/g) {
      $confname = $arg;
      $confname =~ s/-N//; # Strips away the "-N"
      if (length($confname) == 0) {
         $confname = shift(ARGV);
       }
      $confname =~ s/\s//g;
      if($confname =~ /\:/)
        {
          $confname =~ s/\://g;
          print "colons were removed in configuration name: $confname\n";
        }
    }
    elsif($arg =~ "w")
      {
        my $tmp = $arg;
        $tmp =~ s/-w//; # Strips away the "-w"
        if (length($tmp) == 0) {
          $tmp = shift(ARGV);
        }
        $tmp =~ s/\s//g;
        $tmp =~ s/\://g;
        $tmp =~ s/\,//g;
        $weight = $tmp;
        $useweight = 1;
      }
    elsif ($arg =~ "h") {
      $helpwanted = 1;
    }
    elsif ($arg =~ "d") {
      $localdir = 1;
    }
    elsif ($arg =~ "u") {
      $updateDb = 1;
    }
    elsif ($arg =~ "m") {
      $driver = "merge";
      print "option sets mode to $driver\n";
    }
    elsif ($arg =~ "a" && -r "mps.db") {
      $append = 1;
      print "option sets mode to append\n";
    }
    elsif ($arg =~ "-M") {
      $pedeMem = $arg;
      $pedeMem =~ s/-M//; # Strips away the "-M"
      if (length($pedeMem) == 0) { # The memory size must be following param
        $pedeMem = shift(ARGV);
      }
    }

    $optionstring = "$optionstring$arg";
  }
  else {                # parameters not related to options
    $i = $i + 1;
    if ($i eq 1) {
      $batchScript = $arg;
    }
    if ($i eq 2) {
      $cfgTemplate = $arg;
    }
    if ($i eq 3) {
      $infiList = $arg;
    }
    elsif ($i eq 4) {
      $nJobs = $arg;
    }
    elsif ($i eq 5) {
      $class = $arg;
    }
    elsif ($i eq 6) {
      $addFiles = $arg;
    }
    elsif ($i eq 7) {
      $mergeScript = $arg;
    }
    elsif ($i eq 8) {
      $mssDir = $arg;
    }
  }
}

# test input parameters
if ($nJobs eq 0 or $helpwanted != 0 ) {
  print "Usage:\n  mps_setup.pl [options] batchScript cfgTemplate infiList nJobs class[:classMerge] jobname [mergeScript [pool:]mssDir]";
  print "\nKnown options:";
  print "  \n -m          Setup the pede merging job.";
  print "  \n -a          Append jobs to the already existing list.";
  print "  \n -M pedeMem  The memory (MB) to be allocated for pede (min: 1024 MB).";
  print "  \n             If not given, it is evinced from the pede executable name.";
  print "  \n             Finally, it is set 2560 MB if neither of the two are available.";
  print "  \n -N name     Some arbitrary name assigned to the jobs.";
  print "  \n -w value    Statistical weight to be used.";
  print "\n";
  exit 1;
}

unless (-r $batchScript) {
  print "Bad batchScript script name $batchScript\n";
  exit 1;
}
unless (-r $cfgTemplate) {
  print "Bad cfg template file name $cfgTemplate\n";
  exit 1;
}
unless (-r $infiList) {
  print "Bad input list file $infiList\n";
  exit 1;
}
unless (index(" lxplus cmscaf1nh cmscaf1nd cmscaf1nw cmscafspec1nh cmscafspec1nd cmscafspec1nw 8nm 1nh 8nh 1nd 2nd 1nw 2nw "," ".get_class("mille")." ")>-1) {
  print "Bad job class for mille in class '$class'\n";
  exit 1;
}
unless (index(" lxplus cmscaf1nh cmscaf1nd cmscaf1nw cmscafspec1nh cmscafspec1nd cmscafspec1nw 8nm 1nh 8nh 1nd 2nd 1nw 2nw "," ".get_class("pede")." ")>-1) {
  print "Bad job class for pede in class '$class'\n";
  exit 1;
}

if ($driver eq "merge") {
  if ($mergeScript eq "") {
    $mergeScript = $batchScript . "merge";
  }
  unless (-r $mergeScript) {
    print "Bad merge script file name $mergeScript\n";
    exit 1;
  }
}

if ($mssDir ne "") {
  if ($mssDir =~ /:/) { # ':' as delimeter also used in mpedegui.pl
    $mssDirPool = $mssDir;
    $mssDirPool =~ s/:.+?$//; # Remove all that follows ":"

    $mssDir =~ s/^.+?://; # Remove all the precedes ":"
  }

  $testMssDir = `$Mpslib::eos ls -d $mssDir`;
  chomp $testMssDir;
  if ($testMssDir eq "") {
    print "Bad MSS directory name $mssDir\n";
    exit 1;
  }

}

$pedeMemMin = 1024; # Minimum memory allocated for pede: 1024MB=1GB

# Try to guess the memory requirements from the pede executable name.
# 2.5GB is used as default otherwise.
# AP - 23.03.2010
$pedeMemDef = `cat $cfgTemplate | grep "process.AlignmentProducer.algoConfig.pedeSteerer.pedeCommand" | cut -d '#' -f 1 | grep "=" | tail -1 | cut -d '=' -f 2`; # This is the pede executable (full path).
$pedeMemDef = `basename $pedeMemDef`; # This is the pede executable (only the file name, eg "pede_4GB").
$pedeMemDef =~ s/.*_//; # Strip away the first part of the name (eg remains only "4GB").
$pedeMemDef =~ s/GB//; # Strip away "GB" (eg remains only "4").
if ($pedeMemDef =~ /\d/) { # If what remains is a number...
  $pedeMemDef = 1024*$pedeMemDef; # ... convert it into GB.
  if ($pedeMemDef < $pedeMemMin) {$pedeMemDef = $pedeMemMin}; # $pedeMemDef must be >= pedeMemMin.
} else { # Use a default value.
  $pedeMemDef = 1024*2.5;
}

# Allocate memory for the pede job.
# The value specified by the user (-M option) prevails on the one evinced from the executable name.
# AP - 23.03.2010
if ($pedeMem =~ /\D/ || $pedeMem < $pedeMemMin) {
  print "Memory request (".$pedeMem.") non-digit or < ".$pedeMemMin.", use ".$pedeMemDef."\n";
  $pedeMem = $pedeMemDef;
}

# Create the job directories
my $nJobExist="";
if ($append==1 && -d "jobData") {
# Append mode, and "jobData" exists
  $nJobExist = `ls jobData | grep 'job[0-9][0-9][0-9]' | tail -1`;
  $nJobExist =~ s/job//;
}

if ($nJobExist eq "" || $nJobExist <=0 || $nJobExist>999) {
# Delete all
  system "rm -rf jobData";
  system "mkdir jobData";
  $nJobExist = 0;
}

for ($j = 1; $j <= $nJobs; ++$j) {
  $i = $j+$nJobExist;
  $jobdir = sprintf "job%03d",$i;
  print "jobdir $jobdir\n";
  system "mkdir jobData/$jobdir";
}

# build the absolute job directory path (needed by mps_script)
$thePwd = `pwd`;
chomp $thePwd;
$theJobData = "$thePwd/jobData";
print "theJobData= $theJobData \n";

if ($append == 1) {
# save current values
  my $tmpBatchScript = $batchScript;
  my $tmpCfgTemplate = $cfgTemplate;
  my $tmpInfiList = $infiList;
  my $tmpNJobs = $nJobs;
  my $tmpClass = $class;
  my $tmpMergeScript = $mergeScript;
  my $tmpDriver = $driver;

# Read DB file
  read_db();

# check if last job is a merge job
  if (@JOBDIR[$nJobs] eq "jobm") {
# remove the merge job
    pop @JOBDIR;
    pop @JOBID;
    pop @JOBSTATUS;
    pop @JOBNTRY;
    pop @JOBRUNTIME;
    pop @JOBNEVT;
    pop @JOBHOST;
    pop @JOBINCR;
    pop @JOBREMARK;
    pop @JOBSP1;
    pop @JOBSP2;
    pop @JOBSP3;
  }

# Restore variables
  $batchScript = $tmpBatchScript;
  $cfgTemplate = $tmpCfgTemplate;
  $infiList = $tmpInfiList;
  $nJobs = $tmpNJobs;
  $class = $tmpClass;
  $mergeScript = $tmpMergeScript;
  $driver = $tmpDriver;
}


# Create (update) the local database
for ($j = 1; $j <= $nJobs; ++$j) {
  $i=$j+$nJobExist;
  $theJobDir = sprintf "job%03d",$i;
  push @JOBDIR,$theJobDir;
  push @JOBID,0;
  push @JOBSTATUS,"SETUP";
  push @JOBNTRY,0;
  push @JOBRUNTIME,0;
  push @JOBNEVT,0;
  push @JOBHOST,"";
  push @JOBINCR,0;
  push @JOBREMARK,"";
  push @JOBSP1,"";
  if($useweight)
    {
      push @JOBSP2,"$weight";
    }
  else
    {
      push @JOBSP2,"";
    }
  push @JOBSP3,"$confname";
  # create the split card files
  print "mps_split.pl $infiList $j $nJobs >jobData/$theJobDir/theSplit\n";
  system "mps_split.pl $infiList $j $nJobs >jobData/$theJobDir/theSplit";
  if ($?) {
    print "              split failed\n";
    @JOBSTATUS[$i-1] = "FAIL";
  }
  $theIsn = sprintf "%03d",$i;
  print "mps_splice.py $cfgTemplate jobData/$theJobDir/theSplit jobData/$theJobDir/the.py $theIsn\n";
  system "mps_splice.py $cfgTemplate jobData/$theJobDir/theSplit jobData/$theJobDir/the.py $theIsn";
  # create the run script
  print "mps_script.pl $batchScript  jobData/$theJobDir/theScript.sh $theJobData/$theJobDir the.py jobData/$theJobDir/theSplit $theIsn $mssDir $mssDirPool\n";
  system "mps_script.pl $batchScript  jobData/$theJobDir/theScript.sh $theJobData/$theJobDir the.py jobData/$theJobDir/theSplit $theIsn $mssDir $mssDirPool";
}

# create the merge job entry. This is always done. Whether it is used depends on the "merge" option.
$theJobDir = "jobm";
push @JOBDIR,$theJobDir;
push @JOBID,0;
push @JOBSTATUS,"SETUP";
push @JOBNTRY,0;
push @JOBRUNTIME,0;
push @JOBNEVT,0;
push @JOBHOST,"";
push @JOBINCR,0;
push @JOBREMARK,"";
push @JOBSP1,"";
push @JOBSP2,"";
push @JOBSP3,"";

write_db();

# if merge mode, create the directory and set up contents
if ($driver eq "merge") {

  system "rm -rf jobData/jobm";
  system "mkdir jobData/jobm";
  print "Create dir jobData/jobm\n";

  # We want to merge old and new jobs
  my $nJobsMerge = $nJobs+$nJobExist;

  # create  merge job cfg
  print "mps_merge.py -w $cfgTemplate jobData/jobm/alignment_merge.py $theJobData/jobm $nJobsMerge\n";
  system "mps_merge.py -w $cfgTemplate jobData/jobm/alignment_merge.py $theJobData/jobm $nJobsMerge";

  # create merge job script
  print "mps_scriptm.pl $mergeScript jobData/jobm/theScript.sh $theJobData/jobm alignment_merge.py $nJobsMerge $mssDir $mssDirPool\n";
  system "mps_scriptm.pl $mergeScript jobData/jobm/theScript.sh $theJobData/jobm alignment_merge.py $nJobsMerge $mssDir $mssDirPool";
}

# Create a backup of batchScript, cfgTemplate, infiList (and mergeScript)
#   in jobData
$i = `ls jobData | grep 'ScriptsAndCfg[0-9][0-9][0-9]' | tail -1`;
$i =~ s/ScriptsAndCfg//;
$i =~ s/.tar//;
$i++;
$ScriptCfg = sprintf "ScriptsAndCfg%03d",$i;
system "mkdir jobData/$ScriptCfg";
system "cp $batchScript $cfgTemplate $infiList jobData/$ScriptCfg/.";
if ($driver eq "merge") {
  system "cp $mergeScript jobData/$ScriptCfg/.";
}
system "tar -cf jobData/$ScriptCfg.tar jobData/$ScriptCfg";
system "rm -rf jobData/$ScriptCfg";


# Write to DB

write_db();
read_db();
print_memdb();

File: Alignment/MillePedeAlignmentAlgorithm/scripts/mps_setup.py
#!/usr/bin/env python3

from __future__ import print_function
from builtins import range
import os
import re
import sys
import shutil
import tarfile
import argparse
import subprocess
import Alignment.MillePedeAlignmentAlgorithm.mpslib.tools as mps_tools
import Alignment.MillePedeAlignmentAlgorithm.mpslib.Mpslibclass as mpslib


parser = argparse.ArgumentParser(description = "Setup local mps database")
parser.add_argument("-m", "--setup-merge", dest = "setup_merge",
                    action = "store_true", default = False,
                    help = "setup pede merge job")
parser.add_argument("-a", "--append", action = "store_true", default = False,
                    help = "append jobs to existing list")
parser.add_argument("-M", "--memory", type = int, # seems to be obsolete
                    help = "memory (MB) to be allocated for pede")
parser.add_argument("-N", "--name", # remove restrictions on job name?
                    help = ("name to be assigned to the jobs; Whitespaces and "
                            "colons are not allowed"))
parser.add_argument("-w", "--weight", type = float,
                    help = "assign statistical weight")
parser.add_argument("-e", "--max-events", dest = "max_events", type = int,
                    help = "maximum number of events to process")

parser.add_argument("batch_script",
                    help = "path to the mille batch script template")
parser.add_argument("config_template",
                    help = "path to the config template")
parser.add_argument("input_file_list",
                    help = "path to the input file list")
parser.add_argument("n_jobs", type = int,
                    help = "number of jobs assigned to this dataset")
parser.add_argument("job_class",
                    help=("can be any of the normal LSF queues (8nm, 1nh, 8nh, "
                    "1nd, 2nd, 1nw, 2nw), special CAF queues (cmscaf1nh, "
                    "cmscaf1nd, cmscaf1nw) and special CAF pede queues "
                    "(cmscafspec1nh, cmscafspec1nd, cmscafspec1nw); if it "
                    "contains a ':' the part before ':' defines the class for "
                    "mille jobs and the part after defines the pede job class"))
parser.add_argument("job_name",
                    help = "name assigned to batch jobs")
parser.add_argument("merge_script",
                    help = "path to the pede batch script template")
parser.add_argument("mss_dir",
                    help = "name of the mass storage directory")

args = parser.parse_args(sys.argv[1:])


# setup mps database
lib = mpslib.jobdatabase()
lib.batchScript = args.batch_script
lib.cfgTemplate = args.config_template
lib.infiList = args.input_file_list
lib.nJobs = args.n_jobs
lib.classInf = args.job_class
lib.addFiles = args.job_name
lib.driver = "merge" if args.setup_merge else ""
lib.mergeScript = args.merge_script
lib.mssDirPool = ""
lib.mssDir = args.mss_dir
lib.pedeMem = args.memory


if not os.access(args.batch_script, os.R_OK):
    print("Bad 'batch_script' script name", args.batch_script)
    sys.exit(1)

if not os.access(args.config_template, os.R_OK):
    print("Bad 'config_template' file name", args.config_template)
    sys.exit(1)

if not os.access(args.input_file_list, os.R_OK):
    print("Bad input list file", args.input_file_list)
    sys.exit(1)

# ignore 'append' flag if mps database is not yet created
if not os.access("mps.db", os.R_OK): args.append = False

allowed_mille_classes = ("lxplus", "cmscaf1nh", "cmscaf1nd", "cmscaf1nw",
                         "cmscafspec1nh", "cmscafspec1nd", "cmscafspec1nw",
                         "8nm", "1nh", "8nh", "1nd", "2nd", "1nw", "2nw",
                         "cmsexpress","htcondor_cafalca_espresso","htcondor_espresso",
                         "htcondor_cafalca_microcentury","htcondor_microcentury",
                         "htcondor_cafalca_longlunch", "htcondor_longlunch",
                         "htcondor_cafalca_workday", "htcondor_workday",
                         "htcondor_cafalca_tomorrow", "htcondor_tomorrow",
                         "htcondor_cafalca_testmatch", "htcondor_testmatch",
                         "htcondor_cafalca_nextweek", "htcondor_nextweek")
if lib.get_class("mille") not in allowed_mille_classes:
    print("Bad job class for mille in class", args.job_class)
    print("Allowed classes:")
    for mille_class in allowed_mille_classes:
        print(" -", mille_class)
    sys.exit(1)

allowed_pede_classes = ("lxplus", "cmscaf1nh", "cmscaf1nd", "cmscaf1nw",
                        "cmscafspec1nh", "cmscafspec1nd", "cmscafspec1nw",
                        "8nm", "1nh", "8nh", "1nd", "2nd", "1nw", "2nw",
                        "htcondor_bigmem_espresso",
                        "htcondor_bigmem_microcentury",
                        "htcondor_bigmem_longlunch",
                        "htcondor_bigmem_workday",
                        "htcondor_bigmem_tomorrow",
                        "htcondor_bigmem_testmatch",
                        "htcondor_bigmem_nextweek")
if lib.get_class("pede") not in allowed_pede_classes:
    print("Bad job class for pede in class", args.job_class)
    print("Allowed classes:")
    for pede_class in allowed_pede_classes:
        print(" -", pede_class)
    sys.exit(1)

if args.setup_merge:
    if args.merge_script == "":
        args.merge_script = args.batch_script + "merge"
    if not os.access(args.merge_script, os.R_OK):
        print("Bad merge script file name", args.merge_script)
        sys.exit(1)

if args.mss_dir.strip() != "":
    if ":" in args.mss_dir:
        lib.mssDirPool = args.mss_dir.split(":")
        lib.mssDirPool, args.mss_dir = lib.mssDirPool[0], ":".join(lib.mssDirPool[1:])
        lib.mssDir = args.mss_dir

pedeMemMin = 1024 # Minimum memory allocated for pede: 1024MB=1GB

# Try to guess the memory requirements from the pede executable name.
# 2.5GB is used as default otherwise.
# AP - 23.03.2010
cms_process = mps_tools.get_process_object(args.config_template)
pedeMemDef = cms_process.AlignmentProducer.algoConfig.pedeSteerer.pedeCommand.value()
pedeMemDef = os.path.basename(pedeMemDef) # This is the pede executable (only the file name, eg "pede_4GB").
pedeMemDef = pedeMemDef.split("_")[-1]
pedeMemDef = pedeMemDef.replace("GB", "")
try:
    pedeMemDef = 1024*float(pedeMemDef)
    if pedeMemDef < pedeMemMin: pedeMemDef = pedeMemMin # pedeMemDef must be >= pedeMemMin.
except ValueError:
    pedeMemDef = int(1024*2.5)


# Allocate memory for the pede job.
# The value specified by the user (-M option) prevails on the one evinced from the executable name.
# AP - 23.03.2010
if not args.memory or args.memory < pedeMemMin:
    print("Memory request ({}) is < {}, using {}.".format(args.memory, pedeMemMin, pedeMemDef), end=' ')
    lib.pedeMem = args.memory = pedeMemDef

# Create the job directories
nJobExist = 0
if args.append and os.path.isdir("jobData"):
    # Append mode, and "jobData" exists
    jobs = os.listdir("jobData")
    job_regex = re.compile(r"job([0-9]{3})") # should we really restrict it to 3 digits?
    existing_jobs = [job_regex.search(item) for item in jobs]
    existing_jobs = [int(job.group(1)) for job in existing_jobs if job is not None]
    nJobExist = sorted(existing_jobs)[-1]

if nJobExist == 0 or nJobExist <=0 or nJobExist > 999: # quite rude method... -> enforce job number limit earlier?
    # Delete all
    mps_tools.remove_existing_object("jobData")
    os.makedirs("jobData")
    nJobExist = 0;

for j in range(1, args.n_jobs + 1):
    i = j+nJobExist
    jobdir = "job{0:03d}".format(i)
    print("jobdir", jobdir)
    os.makedirs(os.path.join("jobData", jobdir))

# build the absolute job directory path (needed by mps_script)
theJobData = os.path.abspath("jobData")
print("theJobData =", theJobData)

if args.append:
    # save current values
    tmpBatchScript = lib.batchScript
    tmpCfgTemplate = lib.cfgTemplate
    tmpInfiList    = lib.infiList
    tmpNJobs       = lib.nJobs
    tmpClass       = lib.classInf
    tmpMergeScript = lib.mergeScript
    tmpDriver      = lib.driver

    # Read DB file
    lib.read_db()

    # check if last job is a merge job
    if lib.JOBDIR[lib.nJobs] == "jobm":
        # remove the merge job
        lib.JOBDIR.pop()
        lib.JOBID.pop()
        lib.JOBSTATUS.pop()
        lib.JOBNTRY.pop()
        lib.JOBRUNTIME.pop()
        lib.JOBNEVT.pop()
        lib.JOBHOST.pop()
        lib.JOBINCR.pop()
        lib.JOBREMARK.pop()
        lib.JOBSP1.pop()
        lib.JOBSP2.pop()
        lib.JOBSP3.pop()

    # Restore variables
    lib.batchScript = tmpBatchScript
    lib.cfgTemplate = tmpCfgTemplate
    lib.infiList    = tmpInfiList
    lib.nJobs       = tmpNJobs
    lib.classInf    = tmpClass
    lib.mergeScript = tmpMergeScript
    lib.driver      = tmpDriver


# Create (update) the local database
for j in range(1, args.n_jobs + 1):
    i = j+nJobExist
    jobdir = "job{0:03d}".format(i)
    lib.JOBDIR.append(jobdir)
    lib.JOBID.append("")
    lib.JOBSTATUS.append("SETUP")
    lib.JOBNTRY.append(0)
    lib.JOBRUNTIME.append(0)
    lib.JOBNEVT.append(0)
    lib.JOBHOST.append("")
    lib.JOBINCR.append(0)
    lib.JOBREMARK.append("")
    lib.JOBSP1.append("")
    if args.weight is not None:
        lib.JOBSP2.append(str(args.weight))
    else:
        lib.JOBSP2.append("")
    lib.JOBSP3.append(args.name)

    # create the split card files
    cmd = ["mps_split.pl", args.input_file_list,
           str(j if args.max_events is None else 1),
           str(args.n_jobs if args.max_events is None else 1)]
    print(" ".join(cmd)+" > jobData/{}/theSplit".format(jobdir))
    with open("jobData/{}/theSplit".format(jobdir), "w") as f:
        try:
            subprocess.check_call(cmd, stdout = f)
        except subprocess.CalledProcessError:
            print("              split failed")
            lib.JOBSTATUS[i-1] = "FAIL"
    theIsn = "{0:03d}".format(i)

    # create the cfg file
    cmd = ["mps_splice.py", args.config_template,
           "jobData/{}/theSplit".format(jobdir),
           "jobData/{}/the.py".format(jobdir), theIsn]
    if args.max_events is not None:
        chunk_size = int(args.max_events/args.n_jobs)
        event_options = ["--skip-events", str(chunk_size*(j-1))]
        max_events = (args.max_events - (args.n_jobs-1)*chunk_size
                      if j == args.n_jobs    # last job gets the remaining events
                      else chunk_size)
        event_options.extend(["--max-events", str(max_events)])
        cmd.extend(event_options)
    print(" ".join(cmd))
    mps_tools.run_checked(cmd)

    # create the run script
    print("mps_script.pl {}  jobData/{}/theScript.sh {}/{} the.py jobData/{}/theSplit {} {} {}".format(args.batch_script, jobdir, theJobData, jobdir, jobdir, theIsn, args.mss_dir, lib.mssDirPool))
    mps_tools.run_checked(["mps_script.pl", args.batch_script,
                           "jobData/{}/theScript.sh".format(jobdir),
                           os.path.join(theJobData, jobdir), "the.py",
                           "jobData/{}/theSplit".format(jobdir), theIsn,
                           args.mss_dir, lib.mssDirPool])


# create the merge job entry. This is always done. Whether it is used depends on the "merge" option.
jobdir = "jobm";
lib.JOBDIR.append(jobdir)
lib.JOBID.append("")
lib.JOBSTATUS.append("SETUP")
lib.JOBNTRY.append(0)
lib.JOBRUNTIME.append(0)
lib.JOBNEVT.append(0)
lib.JOBHOST.append("")
lib.JOBINCR.append(0)
lib.JOBREMARK.append("")
lib.JOBSP1.append("")
lib.JOBSP2.append("")
lib.JOBSP3.append("")

lib.write_db();

# if merge mode, create the directory and set up contents
if args.setup_merge:
    shutil.rmtree("jobData/jobm", ignore_errors = True)
    os.makedirs("jobData/jobm")
    print("Create dir jobData/jobm")

    # We want to merge old and new jobs
    nJobsMerge = args.n_jobs+nJobExist

    # create  merge job cfg
    print("mps_merge.py -w {} jobData/jobm/alignment_merge.py {}/jobm {}".format(args.config_template, theJobData, nJobsMerge))
    mps_tools.run_checked(["mps_merge.py", "-w", args.config_template,
                           "jobData/jobm/alignment_merge.py",
                           os.path.join(theJobData, "jobm"), str(nJobsMerge)])

    # create merge job script
    print("mps_scriptm.pl {} jobData/jobm/theScript.sh {}/jobm alignment_merge.py {} {} {}".format(args.merge_script, theJobData, nJobsMerge, args.mss_dir, lib.mssDirPool))
    mps_tools.run_checked(["mps_scriptm.pl", args.merge_script,
                           "jobData/jobm/theScript.sh",
                           os.path.join(theJobData, "jobm"),
                           "alignment_merge.py", str(nJobsMerge), args.mss_dir,
                           lib.mssDirPool])


# Create a backup of batchScript, cfgTemplate, infiList (and mergeScript)
#   in jobData
backups = os.listdir("jobData")
bu_regex = re.compile(r"ScriptsAndCfg([0-9]{3})\.tar")
existing_backups = [bu_regex.search(item) for item in backups]
existing_backups = [int(bu.group(1)) for bu in existing_backups if bu is not None]
i = (0 if len(existing_backups) == 0 else sorted(existing_backups)[-1]) + 1
ScriptCfg = "ScriptsAndCfg{0:03d}".format(i)
ScriptCfg = os.path.join("jobData", ScriptCfg)
os.makedirs(ScriptCfg)
for f in (args.batch_script, args.config_template, args.input_file_list):
    shutil.copy2(f, ScriptCfg)
if args.setup_merge:
    shutil.copy2(args.merge_script, ScriptCfg)

with tarfile.open(ScriptCfg+".tar", "w") as tar: tar.add(ScriptCfg)
shutil.rmtree(ScriptCfg)


# Write to DB
lib.write_db();
lib.read_db();
lib.print_memdb();
